### PR TITLE
WEBITOOLS-224: Adds option to include fast scan in deep analysis job

### PR DESCRIPTION
**Pull request by CAST-ANJ**

### CODE REVIEW

**Comment by leon2Meudon92** [comment]
You already have a command that does that
Check this job: https://jenkins15.corp.castsoftware.com/view/AipConsole/view/AipConsole%20Tools/job/AipConsoleTools_V2_TestIntegr_OnboardApplication/

**Comment by CAST-ANJ** [comment]
This is different from that. It is related to the new feature added in the console to include fast scan in deep analysis job It modifies the command to align with the new params of api -> /deep-analysis

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java [hunk 1]
```diff
@@ -71,6 +97,9 @@ public OnboardApplicationDeepAnalysisBuilder(String applicationName) {
     @Override
     protected void performClient(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
         String expandedAppName = environmentVariables.expand(getApplicationName());
+        EnvVars vars = run.getEnvironment(listener);
+        String expandedSourcePath = vars.expand(sourcePath);
```
Comment (leon2Meudon92) [inline aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: environmentVariables.expand(x) does that

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java [hunk 2]
```diff
@@ -71,6 +97,9 @@ public OnboardApplicationDeepAnalysisBuilder(String applicationName) {
...
+        String expandedIncludeFastScan = vars.expand(environmentVariables.get("INCLUDE_FAST_SCAN"));
```
Comment (leon2Meudon92) [inline aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: environmentVariables can also expand

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java [hunk 3]
```diff
@@ -88,6 +117,18 @@ protected void performClient(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath,
                 firstScan = app == null || app.getVersion() == null || StringUtils.isEmpty(app.getVersion().getGuid()) || !app.isOnboarded();
             }
 
+            logger.println("Fast scan included: " + expandedIncludeFastScan);
+
+            String _sourcePath = "";
+            if(Boolean.valueOf(expandedIncludeFastScan)){
+                _sourcePath = uploadService.uploadFileForOnboarding(Paths.get(expandedSourcePath).toFile(), existingAppGuid);
```
Comment (leon2Meudon92) [inline aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: upload only if ANALYZED
If source path is empty what you do
@NotNull constraint should be added

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java [hunk 4]
```diff
@@ -88,6 +117,18 @@ protected void performClient(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath,
...
+                if(app.getVersion().getStatus() == VersionStatus.ANALYZED) {
+                    logger.println("Fast Scan will be done before running deep analysis");
+                } else {
+                    logger.println("Application should be analyzed to include fast scan in deep analysis. Deep analysis will continue without fast scan");
```
Comment (leon2Meudon92) [inline aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: add message resource instead

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java [hunk 5]
```diff
@@ -110,11 +151,15 @@ protected void performClient(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath,
                 moduleType = ModuleGenerationType.fromString(moduleGenerationType);
             }
 
-            applicationService.runDeepAnalysis(existingAppGuid, targetNode, caipVersion, expandedSsnapshotName, moduleType, verbose, jnksLogPollingProvider);
+            applicationService.runDeepAnalysis(existingAppGuid, targetNode, caipVersion
+                    , expandedSsnapshotName, moduleType, verbose
+                    , Boolean.valueOf(expandedIncludeFastScan), _sourcePath, jnksLogPollingProvider);
```
Comment (leon2Meudon92) [inline aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: Boolean.valueOf(expandedIncludeFastScan) should have been converted earlier
boolean b = Boolean.valueOf(expandedIncludeFastScan)
and use b everywhere

File: aip-console-jenkins/src/test/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilderTest.java [hunk 6]
```diff
@@ -43,6 +43,7 @@ public void testOnboardApplicationDeepAnalysisJob() throws Exception {
         Object builtProject = project.getBuildersList().get(0);
         OnboardApplicationDeepAnalysisBuilder expectedResults = new OnboardApplicationDeepAnalysisBuilder(TEST_APP_NAME);
         expectedResults.setSnapshotName("");
+        expectedResults.setSourcePath("");
```
Comment (leon2Meudon92) [inline aip-console-jenkins/src/test/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilderTest.java:46]: Add another test with non empty path

File: aip-console-tools-cli/src/main/java/com/castsoftware/aip/console/tools/commands/OnboardApplicationDeepAnalysisCommand.java [hunk 7]
```diff
@@ -45,6 +47,15 @@ public class OnboardApplicationDeepAnalysisCommand extends BasicCallable {
     @CommandLine.Mixin
     private SharedOptions sharedOptions;
 
+    @CommandLine.Option(names = "--include-fast-scan", description = "If provided, fast-scan will run before deep analysis. The application should already be analysed to enable this option."
+            + " if specified without parameter: ${FALLBACK-VALUE}", fallbackValue = "false")
```
Comment (leon2Meudon92) [inline aip-console-tools-cli/src/main/java/com/castsoftware/aip/console/tools/commands/OnboardApplicationDeepAnalysisCommand.java:51]: fallback should be true

File: aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java [hunk 8]
```diff
@@ -170,6 +171,18 @@ public int deepAnalyze(DeepAnalyzeProperties properties) throws JobServiceExcept
                 app = getApplicationDetails(existingAppGuid);
             }
 
+            boolean includeFastScan = properties.isIncludeFastScan();
+            String sourcePath = "";
+            if(includeFastScan){
+                sourcePath = uploadService.uploadFileAndGetSourcePath(app.getName(), app.getGuid(), properties.getSourcePath());
```
Comment (leon2Meudon92) [inline aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java]: Only perform the upload when required: status==ANALYZED

File: aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java [hunk 9]
```diff
@@ -195,15 +208,18 @@ public int deepAnalyze(DeepAnalyzeProperties properties) throws JobServiceExcept
...
                 log.info("Deep-Analyze done successfully");
             } else {
-                log.error("Deep-Analyze didn't completed successfully.");
+                log.error("Deep-Analyze didn't complete successfully.");
                 return Constants.RETURN_JOB_FAILED;
             }
         } catch (ApplicationServiceException e) {
             return Constants.RETURN_APPLICATION_INFO_MISSING;
+        } catch (UploadException e) {
+            throw new RuntimeException(e);
```
Comment (leon2Meudon92) [inline aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java]: give a precise error code

**Comment by leon2Meudon92** [comment]
They are comments reviews

**Review by leon2Meudon92 — CHANGES_REQUESTED** [review]

File: aip-console-tools-cli/src/main/java/com/castsoftware/aip/console/tools/commands/OnboardApplicationDeepAnalysisCommand.java (hunk 7 above)
Comment (CAST-ANJ) [reply aip-console-tools-cli/src/main/java/com/castsoftware/aip/console/tools/commands/OnboardApplicationDeepAnalysisCommand.java:51]: fallback is false on console side.

**Comment by leon2Meudon92** [comment]
Please don't forget to add a new JENKINS CLI job like the one a showed above

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java (hunk 3 above)
Comment (CAST-ANJ) [reply aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: Changes done

File: aip-console-jenkins/src/test/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilderTest.java (hunk 6 above)
Comment (CAST-ANJ) [reply aip-console-jenkins/src/test/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilderTest.java:46]: Changes done

File: aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java (hunk 8 above)
Comment (CAST-ANJ) [reply aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java]: Changes done

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java (hunk 1 above)
Comment (CAST-ANJ) [reply aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: Changes done

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java (hunk 4 above)
Comment (CAST-ANJ) [reply aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: Changes done

File: aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java (hunk 9 above)
Comment (CAST-ANJ) [reply aip-console-tools-core/src/main/java/com/castsoftware/aip/console/tools/core/services/ApplicationServiceImpl.java]: Changes done

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java (hunk 5 above)
Comment (CAST-ANJ) [reply aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: Changes done

File: aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java (hunk 2 above)
Comment (CAST-ANJ) [reply aip-console-jenkins/src/main/java/io/jenkins/plugins/aipconsole/OnboardApplicationDeepAnalysisBuilder.java]: Changes done
